### PR TITLE
Implement Carroll method for Interior Radiant Heat Exchange

### PR DESCRIPTION
**Pull request by mbadams5**
Pull request overview
---------------------
This PR implements the Carroll method for interior radiant heat exchange instead of the current Hottel's ScriptF method. The Carroll method has linear complexity while the ScriptF method has O(n^2) complexity. This greatly improves runtime (up to 95% runtime reduction) when a zone has a large number of surfaces.

New option added to PerformancePrecisionTradeoffs:
 A2; \field Zone Radiant Exchange Algorithm
      \note Determines which algorithm will be used to solve long wave radiant exchange among surfaces within a zone.
      \type choice
      \key ScriptF
      \key CarrollMRT
      \default ScriptF
### Pull Request Author
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [x] Title of PR should be user-synopsis style (clearly understandable in a standalone changelog context)
 - [x] Label the PR with at least one of: Defect, Refactoring, NewFeature, Performance, and/or DoNoPublish
 - [x] Pull requests that impact EnergyPlus code must also include unit tests to cover enhancement or defect repair
 - [ ] Author should provide a "walkthrough" of relevant code changes using a GitHub code review comment process
 - [ ] If any diffs are expected, author must demonstrate they are justified using plots and descriptions
 - [ ] If changes fix a defect, the fix should be demonstrated in plots and descriptions
 - [ ] If any defect files are updated to a more recent version, upload new versions here or on DevSupport
 - [ ] If IDD requires transition, transition source, rules, ExpandObjects, and IDFs must be updated, and add IDDChange label
 - [ ] If structural output changes, add to output rules file and add OutputChange label
 - [ ] If adding/removing any LaTeX docs or figures, update that document's CMakeLists file dependencies

### Reviewer
This will not be exhaustively relevant to every PR.
 - [x] Perform a Code Review on GitHub
 - [x] If branch is behind develop, merge develop and build locally to check for side effects of the merge
 - [ ] ~If defect, verify by running develop branch and reproducing defect, then running PR and reproducing fix~
 - [x] If feature, test running new feature, try creative ways to break it
 - [x] CI status: all green or justified
 - [x] Check that performance is not impacted (CI Linux results include performance check)
 - [x] Run Unit Test(s) locally
 - [x] Check any new function arguments for performance impacts
 - [x] Verify IDF naming conventions and styles, memos and notes and defaults
 - [x] If new idf included, locally check the err file and other outputs

### CODE REVIEW

File: design/FY2019/Carroll-MRT.md
Comment (mjwitte) [inline design/FY2019/Carroll-MRT.md:12]: So, this is kind of back to the future. If I'm not mistaken, the MRT method was used in BLAST and DOE2.1E. @RKStrand can confirm. The [comments](https://github.com/NREL/EnergyPlus/blob/develop/src/EnergyPlus/HeatBalanceIntRadExchange.cc#L96-L99) at the top of `HeatBalanceIntRadExchange.cc` allude to that.

File: idd/Energy+.idd.in
Comment (mjwitte) [inline idd/Energy+.idd.in]: We have a proliferation of these single-function objects. I'd prefer to see this as a new field in, say, HeatBalanceAlgorithm.
Or a better fit might be the new PerformancePrecisionTradeoffs object since that's the intent here.

File: design/FY2019/Carroll-MRT.md
Comment (RKStrand) [reply design/FY2019/Carroll-MRT.md:12]: Yes, I can confirm that BLAST used the MRT method and so did EnergyPlus early on as it was based on the BLAST heat balance method.  I can't say for DOE-2, but my guess is that it also used MRT or some variation on it.  It doesn't surprise me that this work speeds things up.  The current method for E+ requires potentially a lot of iteration in the interior heat balance.

Comment (nealkruis) [reply design/FY2019/Carroll-MRT.md:12]: @mjwitte @RKStrand I believe there are several competing MRT methods developed by different people:

- Walton (BLAST)
- Seem (TRNSYS, also known as the star-network)
- Carroll (CSE)

What I've heard (from @chipbarnaby), is that Walton's method does not strictly balance heat, whereas Carroll was able to make this work.

I have yet to read Seem's thesis to see how his approach works.

My understanding is that DOE-2 doesn't explicitly model interior radiation exchange (it uses constant combined convection/radation film factors).

File: idd/Energy+.idd.in
Comment (nealkruis) [reply idd/Energy+.idd.in]: @mjwitte I can do that.

Is the PerformancePrecisionTradeoffs object merged in yet? If not, I think it would be best to have these tradeoffs in existing topical settings objects (like HeatBalanceAlgorithm) with a clear document on the various tradeoff options and where to find them. There are already several alternatives that aren't in the PerformancePrecisionTradeoffs that can be included, too (e.g., CTF vs. finite difference, and the air heat balance algorithm).

Comment (mjwitte) [reply idd/Energy+.idd.in]: Yes, it's in v9.2. There is no plan to move existing algorithm options into PerformancePrecisionTradeoffs.

File: design/FY2019/Carroll-MRT.md
Comment (RKStrand) [reply design/FY2019/Carroll-MRT.md:12]: Well, the Walton model without modifications could potentially result in some net radiation imbalances.  To address that, there was an additional term added (RBAL) that resolved that problem.  I believe it was implemented in both BLAST and early EnergyPlus.  So, what was implemented in those programs did balance radiation.

It's been a while since I have looked at Seem's dissertation and what was developed for EnergyPlus from his dissertation did not implement his star network.  However, I think that he did incorporate radiant exchange into his method.  I don't think that is really applicable to EnergyPlus though because the situation is much more complex than was developed in Seem's dissertation from what I remember.

Comment (mjwitte) [reply design/FY2019/Carroll-MRT.md:12]: It would be good to mention these in the NFP.

File: src/EnergyPlus/HeatBalanceIntRadExchange.hh
Comment (mjwitte) [inline src/EnergyPlus/HeatBalanceIntRadExchange.hh]: Looks like the indent got lost here.

File: src/EnergyPlus/HeatBalanceIntRadExchange.cc
Comment (mjwitte) [inline src/EnergyPlus/HeatBalanceIntRadExchange.cc]: `SendSurfTemp` looks wrong here or it's missing an assignment within this loop? As I read it now, `SendSurfTemp` in this loop is left over from the prior for loop at lines 345:373. That variable and others like it really need to be declared within their loops to make sure we aren't reusing an old value somewhere down the line.

File: src/EnergyPlus/HeatBalanceIntRadExchange.cc
Comment (mjwitte) [inline src/EnergyPlus/HeatBalanceIntRadExchange.cc]: It's a little disappointing to see multiple `if (CarrollMethod)` in these loops. But I see there's some significant shared sections. I guess other streamlining of CalcInteriorRadExchange will have to wait for another pass.

File: src/EnergyPlus/HeatBalanceIntRadExchange.cc
Comment (mjwitte) [inline src/EnergyPlus/HeatBalanceIntRadExchange.cc:1965]: If either of the errors in this function trip, does the simulation continue? Seems this should end with a fatal if `errorsFound` is true.

File: src/EnergyPlus/HeatBalanceIntRadExchange.cc
Comment (mjwitte) [inline src/EnergyPlus/HeatBalanceIntRadExchange.cc:1981]: I'm going to commit a one-line change that lets this simple loop vectorize on Windows with Visual Studio. It may already be vectorizing on other platforms already without this? More info [here in the wiki](https://github.com/NREL/EnergyPlus/wiki/Performance-Tuning#auto-parallelization-and-auto-vectorization)

File: tst/EnergyPlus/unit/HeatBalanceIntRadExchange.unit.cc
Comment (mjwitte) [inline tst/EnergyPlus/unit/HeatBalanceIntRadExchange.unit.cc:68]: Good unit test - covers the new functions.

File: src/EnergyPlus/SimulationManager.cc
Comment (mjwitte) [inline src/EnergyPlus/SimulationManager.cc:1179]: So, I see input changes, but no output changes. We let PerformancePrecisionTradeoffs into the code without it generating a lick of output to identify what options are active? Would be great if you could add an eio line to report these (similar to System Convergence Limits, for example). If not here, then post an issue to add it.

File: testfiles/CMakeLists.txt
Comment (mjwitte) [inline testfiles/CMakeLists.txt:144]: Ran this with and without CarrollMRT (unscientifically on my laptop) and saw negligible change in run time. I suppose we don't have a better example file in the test suite. Do you have a test file that you could attach at the top of the PR that illustrates the potential speed-up?

File: doc/engineering-reference/src/surface-heat-balance-manager-processes/inside-heat-balance.tex
Comment (mjwitte) [inline doc/engineering-reference/src/surface-heat-balance-manager-processes/inside-heat-balance.tex]: 1980 and 1980a look like the same paper, just formatted a little differently.

**Review by mjwitte — COMMENTED** [review]
Getting close. See comments below. Checked in a small tweak to the code and some doc cleanups (easier to just fix than to list here).

File: tst/EnergyPlus/unit/HeatBalanceIntRadExchange.unit.cc
Comment (mjwitte) [inline tst/EnergyPlus/unit/HeatBalanceIntRadExchange.unit.cc]: This test fails locally on windows, because of mismatched line endings. Need to use a delimited string for the comparison error message, like [here](https://github.com/NREL/EnergyPlus/blob/1bd7cc2767b6e38c4abc49b77578669026638f88/tst/EnergyPlus/unit/HeatBalanceIntRadExchange.unit.cc#L119-L127).

**Comment by nealkruis** [comment]
@mjwitte @RKStrand This is ready for another review.

File: src/EnergyPlus/HeatBalanceIntRadExchange.cc
Comment (mjwitte) [inline src/EnergyPlus/HeatBalanceIntRadExchange.cc]: Good to clarify *rec*_surface

**Review by mjwitte — APPROVED** [review]
Comments have been addressed.
#7646 posted to add eio output for PerformancePrecisionTradeoffs.
Ran the new example file with and without the CarrollMRT option, annual source energy differs by about 1.5%.
Merged in develop one more time to check performance.

**Comment by mjwitte** [comment]
@RKStrand Do you want to review before this is merged?

**Comment by RKStrand** [comment]
@mjwitte I'm not sure when I will get around to doing this (finals start next week) and having looked around at the code once already I trust your review.  I don't want to hold this up.

File: src/EnergyPlus/HeatBalanceIntRadExchange.cc
Comment (mjwitte) [inline src/EnergyPlus/HeatBalanceIntRadExchange.cc]: These lines appear to be identical to the logic in lines 348:360. Why not save SendSurfTemp in an array and just grab it here instead of going through all these gymnastics all over again? Seems that would improve the CMRT speed.

And then we're doing all the same stuff again for RecSurfs in 394:420? This one gets emissivities too, but seems we could do that up in 348:360, save arrays of temps and emissivities and get rid of all of these ifs.

**Review by mjwitte — COMMENTED** [review]
@nealkruis I know a performance slowdown of 0.49% is small, it seems we can do better than that here. `CalcInteriorRadExchange` is a known hot spot that gets called many times. There are four `if (CarrollMethod)` blocks, two of which are inside surface loops. It seems small, but these add up. 

Looking for ways to reduce these extra ifs. The brute force way would be two separate `CalcInteriorRadExchange` functions, one for CMRT and one for ScriptF. I realize that could result is signficant chunks of duplicate code. So, then we could break those out into some smaller functions? Just thinking out loud here.

File: src/EnergyPlus/HeatBalanceIntRadExchange.cc
Comment (amirroth) [reply src/EnergyPlus/HeatBalanceIntRadExchange.cc]: Here is something else to consider. There is a possibility that we will want to run both ScriptF and CarrollMRT within the same simulation. For the same enclosure. In the same timestep. My guess is that the difference between ScriptF and CarrollMRT is not uniform, but rather geometry dependent. I can see an optimization where we run both for all enclosures for a few time steps and then choose the algorithm on an enclosure basis based on whether the two algorithms produce similar results within threshold (CarrollMRT) or not (ScriptF).

**Comment by chipbarnaby** [comment]
Don't think this is possible.  The Carroll factors have the property that
they result in first-law-balanced radiant exchange using a single effective
radiant temperature for the space.  Carroll in his papers showed that the
resulting radiant fluxes are very good approximations to the "real" values
(e.g. ScriptF), but not identical.  They are exact for cubic spaces and
surprisingly close for other shapes, even those that are pretty non-cubic.

Running both at the same time would require parallel simulations, since
different radiant exchange would result in different surface temp history
-- the whole state of the space would be (slightly) different.

**Comment by amirroth** [comment]
Running CarrollMRT, writing the surface temperatures (or net heat fluxes) into a temporary array, comparing with ScriptF and after a few iterations choosing whether to use CarrollMRT full time is "not possible"?

**Comment by chipbarnaby** [comment]
Yes, that would be possible.  But the conduction history would be different
etc.

If the question is whether the Carroll approximation is good enough for a
particular space shape, I think that could be assessed by comparing view
factors.  That is what Carroll did in his original papers that led him to
assert the method is surprisingly accurate.  I am supposed to be searching
for the papers (I think I have hard copies) ... his methodology would be a
good starting point for putting together an algorithm for on-the-fly
scriptF vs. Carroll determination.

**Comment by RKStrand** [comment]
I think that writing results into different arrays is possible (it would have to be done for both surface temperature and heat flux histories for all surfaces), but it's probably going to be a bit messy (in a section of code that already isn't a shining example of easy to follow).  The question is how many time steps would you need to go before you had an accurate assessment on whether or not the results are similar enough.  I think Chip's idea that checking the view factors might be a little less painful if he can find what he is looking for in the literature.

One other thing--whatever method would need to perhaps get a sense for how long the simulation is going to be.  It might not be worth switching to a faster method after a period of time if the user is only doing a design day or two?

File: src/EnergyPlus/HeatBalanceIntRadExchange.cc
Comment (nealkruis) [reply src/EnergyPlus/HeatBalanceIntRadExchange.cc]: @mjwitte I like your suggestion. Let me see if I can clean it up a bit.

It's always a question of where to stop. This whole file could probably use a refactor/redesign, especially if we are moving towards enclosures as a concept. We could make an `Enclosure` class that contains many of these methods broken out into a cleaner design to avoid repetition.

**Comment by amirroth** [comment]
Chip suggested that we look at view factors instead of surface heat-fluxes/temps, and that's probably a cleaner idea.

**Comment by nealkruis** [comment]
@mjwitte I think the remaining diffs are related to the fact that receiving and sending surface temperatures were not being calculated consistently before when a complex glazing system had between-glass shading/blinds.

In these cases, the sending surface temperature was calculated as [`surface_window.EffInsSurfTemp`](https://github.com/NREL/EnergyPlus/blob/d0361cca92e8ea80e54da7e21067ce550ae5a391/src/EnergyPlus/HeatBalanceIntRadExchange.cc#L340) and the receiving surfaces was calculated as [`SurfaceTemp(RecSurfNum)`](https://github.com/NREL/EnergyPlus/blob/d0361cca92e8ea80e54da7e21067ce550ae5a391/src/EnergyPlus/HeatBalanceIntRadExchange.cc#L387).

It's not clear what it should be. In this branch I'm using `SurfaceTemp(SurfNum)`.

**Comment by mjwitte** [comment]
@nealkruis I've made a couple of commits that eliminate the complex glazing diffs and give a slight speed improvement now. If you agree with these changes, I'll merge.

**Comment by Myoldmopar** [comment]
This _does_ look clean from a CI perspective, though the branch is now kinda out-of-date.  Is it worth just pulling develop in to get final results before merge?  I'm not saying we need to, just wondering.

**Comment by mjwitte** [comment]
@nealkruis This will merge after CI comes back all green and pretty. If you want to review my changes before this drops in, now's your chance.

**Comment by mjwitte** [comment]
@Myoldmopar Is something amiss on Mac?

**Comment by Myoldmopar** [comment]
Yes, something looks to have gone awry.  Disregard for the moment and continue merging as needed.  I'll take a look.

**Comment by Myoldmopar** [comment]
I took the Mac down for a full reboot and got it going again.  It was working great recently so hopefully that's all it needed.

**Pull request merged**
